cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,312 @@
+	// set log level
+	ctx := opts.LoggingFlagOpts.InitializeLogger(command.Context())
+	// core process
+	switch opts.pluginSourceType {
+	case notationplugin.PluginSourceTypeFile:
+		if err := installPlugin(ctx, opts.pluginSource, opts.inputChecksum, opts.force); err != nil {
+			return fmt.Errorf("plugin installation failed: %w", err)
+		}
+		return nil
+	case notationplugin.PluginSourceTypeURL:
+		if opts.inputChecksum == "" {
+			return errors.New("installing from URL requires non-empty SHA256 checksum of the plugin source")
+		}
+		pluginURL, err := url.Parse(opts.pluginSource)
+		if err != nil {
+			return fmt.Errorf("failed to parse plugin download URL %s with error: %w", pluginURL, err)
+		}
+		if pluginURL.Scheme != "https" {
+			return fmt.Errorf("failed to download plugin from URL: only the HTTPS scheme is supported, but got %s", pluginURL.Scheme)
+		}
+		tmpFile, err := os.CreateTemp("", notationPluginDownloadTmpFile)
+		if err != nil {
+			return fmt.Errorf("failed to create notationPluginDownloadTmpFile: %w", err)
+		}
+		defer os.Remove(tmpFile.Name())
+		defer tmpFile.Close()
+		err = notationplugin.DownloadPluginFromURL(ctx, opts.pluginSource, tmpFile)
+		if err != nil {
+			return fmt.Errorf("failed to download plugin from URL %s with error: %w", opts.pluginSource, err)
+		}
+		if err := installPlugin(ctx, tmpFile.Name(), opts.inputChecksum, opts.force); err != nil {
+			return fmt.Errorf("plugin installation failed: %w", err)
+		}
+		return nil
+	default:
+		return errors.New("plugin installation failed: unknown plugin source type")
+	}
+}
+
+// installPlugin installs the plugin given plugin source path
+func installPlugin(ctx context.Context, inputPath string, inputChecksum string, force bool) error {
+	// sanity check
+	inputFileStat, err := os.Stat(inputPath)
+	if err != nil {
+		return err
+	}
+	if !inputFileStat.Mode().IsRegular() {
+		return fmt.Errorf("%s is not a valid file", inputPath)
+	}
+	// checksum check
+	if inputChecksum != "" {
+		if err := osutil.ValidateSHA256Sum(inputPath, inputChecksum); err != nil {
+			return err
+		}
+	}
+	// install the plugin based on file type
+	fileType, err := osutil.DetectFileType(inputPath)
+	if err != nil {
+		return err
+	}
+	switch fileType {
+	case notationplugin.MediaTypeZip:
+		rc, err := zip.OpenReader(inputPath)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		return installPluginFromFS(ctx, rc, force)
+	case notationplugin.MediaTypeGzip:
+		// when file is gzip, required to be tar
+		return installPluginFromTarGz(ctx, inputPath, force)
+	default:
+		// input file is not in zip or gzip, try install directly
+		installOpts := plugin.CLIInstallOptions{
+			PluginPath: inputPath,
+			Overwrite:  force,
+		}
+		return installPluginWithOptions(ctx, installOpts)
+	}
+}
+
+// installPluginFromFS extracts, validates and installs the plugin files
+// from a fs.FS
+//
+// Note: zip.ReadCloser implments fs.FS
```

Comment:
info: just FYI that tar can also be mapped to `fs.FS`: https://github.com/oras-project/oras-go/blob/main/internal/fs/tarfs/tarfs.go but gzipped tar cannot.